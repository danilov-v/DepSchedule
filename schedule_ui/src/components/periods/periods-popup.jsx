import React from "react";
import PropTypes from "prop-types";
import {
  Button,
  Form,
  FormGroup,
  FormFeedback,
  Label,
  Input,
  Modal,
  ModalHeader,
  ModalBody,
  ModalFooter,
} from "reactstrap";
import DatePicker from "react-datepicker";
import classnames from "classnames";
import { isDate } from "date-fns";
import { useForm } from "helpers/hooks/useForm";
import { NotificationManager } from "helpers/notification-manager";
import {
  SUCCESS_PERIOD_NOTIFICATION_DATA,
  FAILED_PERIOD_NOTIFICATION_DATA,
} from "constants/notifications";

const validatePeriodForm = values => {
  let errors = {};

  if (!isDate(values.startDate) && !isDate(values.endDate))
    errors["dates"] = "Необходимо выбрать дату периода";

  if (!isDate(values.startDate))
    errors["startDate"] = "Необходимо выбрать дату начала периода";

  if (!isDate(values.endDate))
    errors["endDate"] = "Необходимо выбрать дату конца периода";

  if (!values.name || !values.name.trim())
    errors["name"] = "Назвние должно содержать минимум 1 символ";

  return errors;
};

export function PeriodsPopup({
  isOpen,
  toggle,
  type,
  onPeriodSubmit,
  defaultFormData,
}) {
  const {
    onChange,
    onSubmit,
    errors,
    errorsShown,
    setErrors,
    values,
  } = useForm(submitForm, defaultFormData, validatePeriodForm);

  const onInputChange = ({ target }) =>
    onChange({ [target.name]: target.value });

  const handleChangeDate = fieldName => date => onChange({ [fieldName]: date });

  const handleError = ({ code, userMessage, devMessage }) => {
    NotificationManager.fire(FAILED_PERIOD_NOTIFICATION_DATA);
    console.log(devMessage);

    switch (code) {
      case "INTERSECTION_OF_PERIODS":
        setErrors({ dates: userMessage });
        break;
      default:
        return;
    }
  };

  async function submitForm() {
    try {
      await onPeriodSubmit(values);
      toggle();

      NotificationManager.fire(SUCCESS_PERIOD_NOTIFICATION_DATA);
    } catch (e) {
      handleError(e);
    }
  }
  const { name, startDate, endDate } = values;

  return (
    <Modal isOpen={isOpen} toggle={toggle}>
      <Form className="p-3">
        <ModalHeader toggle={toggle}>
          {type === "create" ? "Создание" : "Редакитрование"} Периоды
        </ModalHeader>
        <ModalBody>
          <FormGroup>
            <Label for="description">Имя Периода</Label>
            <Input
              type="text"
              rows="6"
              name="name"
              placeholder="Период..."
              value={name}
              onChange={onInputChange}
              invalid={!!errors["name"] && errorsShown}
            />
            <FormFeedback>{errors["name"]}</FormFeedback>
          </FormGroup>
          <Label for="unitParent">Длительность События</Label>
          <FormGroup>
            <div className="d-flex">
              <DatePicker
                selected={startDate}
                selectsStart
                startDate={startDate}
                endDate={endDate}
                maxDate={endDate}
                onChange={handleChangeDate("startDate")}
                locale="ru"
                dateFormat="dd/MM/yyyy"
                placeholderText="15/09/2019"
                className={classnames("form-control", {
                  "is-invalid":
                    errorsShown && (errors["dates"] || errors["startDate"]),
                })}
              />
              <DatePicker
                selected={endDate}
                selectsEnd
                startDate={endDate}
                endDate={endDate}
                onChange={handleChangeDate("endDate")}
                minDate={startDate}
                locale="ru"
                dateFormat="dd/MM/yyyy"
                placeholderText="04/12/1995"
                className={classnames("form-control", {
                  "is-invalid":
                    errorsShown && (errors["dates"] || errors["endDate"]),
                })}
              />
            </div>
            <FormFeedback
              className={classnames({
                "d-block":
                  errorsShown &&
                  (errors["dates"] || errors["startDate"] || errors["endDate"]),
              })}
            >
              {errors["dates"] || errors["startDate"] || errors["endDate"]}
            </FormFeedback>
          </FormGroup>
        </ModalBody>
        <ModalFooter>
          <Button
            type="submit"
            color="success"
            className="mr-3"
            onClick={onSubmit}
          >
            {type === "create" ? "Создать" : "Обновить"}
          </Button>
          <Button color="primary" onClick={toggle}>
            Закрыть
          </Button>
        </ModalFooter>
      </Form>
    </Modal>
  );
}

PeriodsPopup.propTypes = {
  toggle: PropTypes.func.isRequired,
  isOpen: PropTypes.bool,
  type: PropTypes.string,
  onPeriodSubmit: PropTypes.func,
  defaultFormData: PropTypes.shape({
    name: PropTypes.string,
    startDate: PropTypes.instanceOf(Date),
    endDate: PropTypes.instanceOf(Date),
  }),
};

PeriodsPopup.defaultProps = {
  isOpen: false,
};
