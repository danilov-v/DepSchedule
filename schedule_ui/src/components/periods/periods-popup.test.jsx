import React from "react";
import shallow from "enzyme/shallow";
import { PeriodsPopup, validatePeriodForm } from "./periods-popup";

const props = {
  toggle: () => {},
  onPeriodSubmit: () => {},
  isOpen: true,
  defaultFormData: {
    name: "test",
    startDate: null,
    endDate: null,
  },
};

describe("periods-popup.jsx", () => {
  describe("<PeriodsPopup />", () => {
    test("should render and match snapshot with create type", () => {
      const component = shallow(<PeriodsPopup {...props} type="create" />);

      expect(component).toMatchSnapshot();
    });

    test("should render and match snapshot with edit type", () => {
      const component = shallow(<PeriodsPopup {...props} type="edit" />);

      expect(component).toMatchSnapshot();
    });

    test("should return map of all validation errors", () => {
      expect(
        validatePeriodForm({
          startDate: null,
          endDate: null,
          name: "",
        })
      ).toEqual({
        dates: "Необходимо выбрать дату периода",
        startDate: "Необходимо выбрать дату начала периода",
        endDate: "Необходимо выбрать дату конца периода",
        name: "Назвние должно содержать минимум 1 символ",
      });
    });

    test("should return name error", () => {
      expect(
        validatePeriodForm({
          startDate: new Date(),
          endDate: new Date(),
          name: "   ",
        })
      ).toEqual({
        name: "Назвние должно содержать минимум 1 символ",
      });
    });

    test("should return emty error object", () => {
      expect(
        validatePeriodForm({
          startDate: new Date(),
          endDate: new Date(),
          name: "asd",
        })
      ).toEqual({});
    });
  });
});
