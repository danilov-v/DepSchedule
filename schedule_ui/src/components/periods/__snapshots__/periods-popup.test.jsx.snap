// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`periods-popup.jsx <PeriodsPopup /> should render and match snapshot with create type 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={false}
  fade={true}
  isOpen={true}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  returnFocusAfterClose={true}
  role="dialog"
  scrollable={false}
  toggle={[Function]}
  unmountOnClose={true}
  zIndex={1050}
>
  <Form
    className="p-3"
    tag="form"
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      Создание
       Периода
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <FormGroup
        tag="div"
      >
        <Label
          for="description"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          Имя Периода
        </Label>
        <Input
          invalid={false}
          name="name"
          onChange={[Function]}
          placeholder="Период..."
          rows="6"
          type="text"
          value="test"
        />
        <FormFeedback
          tag="div"
        />
      </FormGroup>
      <Label
        for="unitParent"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Длительность События
      </Label>
      <FormGroup
        tag="div"
      >
        <div
          className="d-flex"
        >
          <DatePicker
            allowSameDay={false}
            className="form-control"
            dateFormat="dd/MM/yyyy"
            dateFormatCalendar="LLLL yyyy"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            endDate={null}
            inlineFocusSelectedMonth={false}
            locale="ru"
            maxDate={null}
            monthsShown={1}
            nextMonthButtonLabel="Next month"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onInputClick={[Function]}
            onInputError={[Function]}
            onKeyDown={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            onYearChange={[Function]}
            placeholderText="13/04/1995"
            preventOpenOnFocus={false}
            previousMonthButtonLabel="Previous Month"
            readOnly={false}
            renderDayContents={[Function]}
            selected={null}
            selectsStart={true}
            shouldCloseOnSelect={true}
            showMonthYearPicker={false}
            showTimeInput={false}
            showTimeSelect={false}
            startDate={null}
            strictParsing={false}
            timeCaption="Time"
            timeInputLabel="Time"
            timeIntervals={30}
            withPortal={false}
          />
          <DatePicker
            allowSameDay={false}
            className="form-control"
            dateFormat="dd/MM/yyyy"
            dateFormatCalendar="LLLL yyyy"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            endDate={null}
            inlineFocusSelectedMonth={false}
            locale="ru"
            minDate={null}
            monthsShown={1}
            nextMonthButtonLabel="Next month"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onInputClick={[Function]}
            onInputError={[Function]}
            onKeyDown={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            onYearChange={[Function]}
            placeholderText="04/12/1995"
            preventOpenOnFocus={false}
            previousMonthButtonLabel="Previous Month"
            readOnly={false}
            renderDayContents={[Function]}
            selected={null}
            selectsEnd={true}
            shouldCloseOnSelect={true}
            showMonthYearPicker={false}
            showTimeInput={false}
            showTimeSelect={false}
            startDate={null}
            strictParsing={false}
            timeCaption="Time"
            timeInputLabel="Time"
            timeIntervals={30}
            withPortal={false}
          />
        </div>
        <FormFeedback
          className=""
          tag="div"
        />
      </FormGroup>
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        className="mr-3"
        color="success"
        onClick={[Function]}
        tag="button"
        type="submit"
      >
        Создать
      </Button>
      <Button
        color="primary"
        onClick={[Function]}
        tag="button"
      >
        Закрыть
      </Button>
    </ModalFooter>
  </Form>
</Modal>
`;

exports[`periods-popup.jsx <PeriodsPopup /> should render and match snapshot with edit type 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={false}
  fade={true}
  isOpen={true}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  returnFocusAfterClose={true}
  role="dialog"
  scrollable={false}
  toggle={[Function]}
  unmountOnClose={true}
  zIndex={1050}
>
  <Form
    className="p-3"
    tag="form"
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      Редакитрование
       Периода
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <FormGroup
        tag="div"
      >
        <Label
          for="description"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          Имя Периода
        </Label>
        <Input
          invalid={false}
          name="name"
          onChange={[Function]}
          placeholder="Период..."
          rows="6"
          type="text"
          value="test"
        />
        <FormFeedback
          tag="div"
        />
      </FormGroup>
      <Label
        for="unitParent"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Длительность События
      </Label>
      <FormGroup
        tag="div"
      >
        <div
          className="d-flex"
        >
          <DatePicker
            allowSameDay={false}
            className="form-control"
            dateFormat="dd/MM/yyyy"
            dateFormatCalendar="LLLL yyyy"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            endDate={null}
            inlineFocusSelectedMonth={false}
            locale="ru"
            maxDate={null}
            monthsShown={1}
            nextMonthButtonLabel="Next month"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onInputClick={[Function]}
            onInputError={[Function]}
            onKeyDown={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            onYearChange={[Function]}
            placeholderText="13/04/1995"
            preventOpenOnFocus={false}
            previousMonthButtonLabel="Previous Month"
            readOnly={false}
            renderDayContents={[Function]}
            selected={null}
            selectsStart={true}
            shouldCloseOnSelect={true}
            showMonthYearPicker={false}
            showTimeInput={false}
            showTimeSelect={false}
            startDate={null}
            strictParsing={false}
            timeCaption="Time"
            timeInputLabel="Time"
            timeIntervals={30}
            withPortal={false}
          />
          <DatePicker
            allowSameDay={false}
            className="form-control"
            dateFormat="dd/MM/yyyy"
            dateFormatCalendar="LLLL yyyy"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            endDate={null}
            inlineFocusSelectedMonth={false}
            locale="ru"
            minDate={null}
            monthsShown={1}
            nextMonthButtonLabel="Next month"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onInputClick={[Function]}
            onInputError={[Function]}
            onKeyDown={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            onYearChange={[Function]}
            placeholderText="04/12/1995"
            preventOpenOnFocus={false}
            previousMonthButtonLabel="Previous Month"
            readOnly={false}
            renderDayContents={[Function]}
            selected={null}
            selectsEnd={true}
            shouldCloseOnSelect={true}
            showMonthYearPicker={false}
            showTimeInput={false}
            showTimeSelect={false}
            startDate={null}
            strictParsing={false}
            timeCaption="Time"
            timeInputLabel="Time"
            timeIntervals={30}
            withPortal={false}
          />
        </div>
        <FormFeedback
          className=""
          tag="div"
        />
      </FormGroup>
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        className="mr-3"
        color="success"
        onClick={[Function]}
        tag="button"
        type="submit"
      >
        Обновить
      </Button>
      <Button
        color="primary"
        onClick={[Function]}
        tag="button"
      >
        Закрыть
      </Button>
    </ModalFooter>
  </Form>
</Modal>
`;
